### ISSUE TITLE
Hyper-parameter Settings In Paper vs Code

### ISSUE THREAD

**Issue by jhlau** [facebook/NAMAS#8]
Hi, I have a quick question regarding hyper-parameter settings mentioned in the paper vs. those set in train_model.sh

In the paper:
D = 200, H = 400, C = 5, L = 3, and Q = 2

In train_model.sh:
 -miniBatchSize  64 \
 -embeddingDim  64 \
 -bowDim  200 \
 -hiddenSize  64 \
 -epochs  20 \
 -learningRate 0.1 \
 -validArticleDir  $OUT_DIR/valid.filter/article/ \
 -validTitleDir  $OUT_DIR/valid.filter/title/ \
 -window  $WINDOW \
 -printEvery   100 \
 -encoderModel  "attenbow" \
 -attenPool  5

Are these two sets of hyper-parameters the same? I thought embeddingDim (decoder embedding dimension) has the same size as bowDim (encoder embedding dimension). Also, shouldn't hiddenSize be 400 (H=400)? I am also a little confused as to what C, L and Q map to in the code. Any clarification would be appreciated. Cheers.

**Comment by falcondai** [comment]
this does seem to reveal some of the inconsistency between the codebase and the published article. you are right, some of these have different values than the published values 😕

`bowDim` (encoder size or encoder embedding dim) seems to be a hyperparameter separate from the `hiddenSize`: the encoder transforms the variable length article input into fixed dim context vector, `H` in the discussion of encoders in the paper. and in `nnlm.lua` it is considered as an additional constant `P`
- D = embeddingDim
- H = hiddenSize 
- Q = attenPool 
- C = window
- L = 3, number of layers in convolutional encoder, is hardcoded in https://github.com/facebook/NAMAS/blame/master/summary/encoder.lua#L79-L80 and it is not used in attention encoder

according to https://github.com/facebook/NAMAS/blame/master/summary/nnlm.lua#L50 and https://github.com/facebook/NAMAS/blame/master/summary/encoder.lua#L156

**Comment by jhlau** [comment]
Right. So is it correct to say that these should be the hyper-parameter values in the code (to match the paper's):
-miniBatchSize 64 \
-embeddingDim 200 \
-bowDim 200 \
-hiddenSize 400 \
-epochs 15 \
-learningRate 0.05 \
-validArticleDir $OUT_DIR/valid.filter/article/ \
-validTitleDir $OUT_DIR/valid.filter/title/ \
-window 5 \
-printEvery 100 \
-encoderModel "attenbow" \
-attenPool 2

**Comment by falcondai** [comment]
These does look more faithful to the hyperparameters described in the paper. But both in terms of perplexity (27.74 vs 27.1 in the paper) and test examples generated, I found the default setting in code to be similar to the published figures. I would be interested to see the results using this particular setting.

Actually, aside from faithfulness, I would suggest using a bigger `miniBatchSize` if you have a Titan X. The default setting with `miniBatchSize = 64` would under-utilize the 12G video memory on Titan X.
